File: 2023/day-02/main2.dart
import 'dart:io';

// Class to represent the count of cubes of each color
class CubeCount {
  int red;
  int green;
  int blue;

  CubeCount({required this.red, required this.green, required this.blue});

  // Finds the maximum number of cubes needed for each color from a list
  static CubeCount max(List<CubeCount> counts) {
    int maxRed = 0, maxGreen = 0, maxBlue = 0;
    for (var count in counts) {
      if (count.red > maxRed) maxRed = count.red;
      if (count.green > maxGreen) maxGreen = count.green;
      if (count.blue > maxBlue) maxBlue = count.blue;
    }
    return CubeCount(red: maxRed, green: maxGreen, blue: maxBlue);
  }
}

// Function to parse game data from a line
List<CubeCount> parseGameData(String line) {
  var subsetsStr = line.split(': ')[1];
  List<CubeCount> gameData = [];

  for (var subset in subsetsStr.split('; ')) {
    var subsetCounts = {'red': 0, 'green': 0, 'blue': 0};
    for (var match in RegExp(r'(\d+) (\w+)').allMatches(subset)) {
      var count = int.parse(match.group(1)!);
      var color = match.group(2)!;
      subsetCounts[color] = subsetCounts[color]! + count;
    }
    gameData.add(CubeCount(
      red: subsetCounts['red']!,
      green: subsetCounts['green']!,
      blue: subsetCounts['blue']!,
    ));
  }

  return gameData;
}

// Function to calculate the power of a given set of cubes
int calculatePower(int red, int green, int blue) {
  return red * green * blue;
}

// Function to find the total power of the minimum sets of cubes
Future<int> findTotalPower(String filePath) async {
  final file = File(filePath);
  if (!await file.exists()) {
    throw Exception('File not found: $filePath');
  }

  final lines = await file.readAsLines();
  int totalPower = 0;

  for (var line in lines) {
    if (line.isNotEmpty) {
      var gameData = parseGameData(line);
      var minCubes = CubeCount.max(gameData);
      var power = calculatePower(minCubes.red, minCubes.green, minCubes.blue);
      totalPower += power;
    }
  }

  return totalPower;
}

void main() async {
  const filePath = 'input.txt';
  try {
    final totalPower = await findTotalPower(filePath);
    print('Sum of powers of the minimum sets of cubes: $totalPower');
  } catch (e) {
    print('Error: $e');
  }
}

File: 2023/day-02/main2.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

type CubeCountPart2 struct {
	Red   int
	Green int
	Blue  int
}

func ParseGameDataPart2(filePath string) (map[int][]CubeCountPart2, error) {
	games := make(map[int][]CubeCountPart2)

	file, err := os.Open(filePath)
	if err != nil {
		return nil, err
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		line := scanner.Text()
		if line == "" {
			continue
		}

		parts := strings.Split(line, ": ")
		gameIDStr := strings.Split(parts[0], " ")[1]
		gameID, err := strconv.Atoi(gameIDStr)
		if err != nil {
			return nil, err
		}

		subsetsStr := strings.Split(parts[1], "; ")
		var gameData []CubeCountPart2

		for _, subsetStr := range subsetsStr {
			cubesStr := strings.Split(subsetStr, ", ")
			cubeCount := CubeCountPart2{}
			for _, cubeStr := range cubesStr {
				countColor := strings.Split(cubeStr, " ")
				count, err := strconv.Atoi(countColor[0])
				if err != nil {
					return nil, err
				}
				color := countColor[1]

				switch color {
				case "red":
					cubeCount.Red += count
				case "green":
					cubeCount.Green += count
				case "blue":
					cubeCount.Blue += count
				}
			}
			gameData = append(gameData, cubeCount)
		}

		games[gameID] = gameData
	}

	if err := scanner.Err(); err != nil {
		return nil, err
	}

	return games, nil
}

func CalculateMinimumCubesPart2(gameData []CubeCountPart2) (int, int, int) {
	minRed, minGreen, minBlue := 0, 0, 0

	for _, subset := range gameData {
		if subset.Red > minRed {
			minRed = subset.Red
		}
		if subset.Green > minGreen {
			minGreen = subset.Green
		}
		if subset.Blue > minBlue {
			minBlue = subset.Blue
		}
	}

	return minRed, minGreen, minBlue
}

func CalculatePowerPart2(red, green, blue int) int {
	return red * green * blue
}

func FindTotalPowerPart2(filePath string) (int, error) {
	games, err := ParseGameDataPart2(filePath)
	if err != nil {
		return 0, err
	}

	totalPower := 0

	for _, gameData := range games {
		minRed, minGreen, minBlue := CalculateMinimumCubesPart2(gameData)
		power := CalculatePowerPart2(minRed, minGreen, minBlue)
		totalPower += power
	}

	return totalPower, nil
}

func main() {
	filePath := "input.txt"
	totalPower, err := FindTotalPowerPart2(filePath)

	if err != nil {
		fmt.Println("Error:", err)
		return
	}

	fmt.Printf("Total power: %d\n", totalPower)
}
